src/main/java/io/pivotal/cfapp/task/AppDetailTask.java

```diff
@@ -38,26 +44,35 @@
 
 @Slf4j
 @Component
-public class AppDetailTask implements ApplicationListener<SpacesRetrievedEvent> {
+public class AppDetailTask implements ApplicationListener<ApplicationEvent> {
 
     private final PasSettings settings;
+    private final PivnetSettings pivnetSettings;
     private final DefaultCloudFoundryOperations opsClient;
     private final AppDetailService appDetailsService;
     private final EventsService eventsService;
     private final BuildpacksCache buildpacksCache;
     private final StacksCache stacksCache;
     private final ApplicationEventPublisher publisher;
+    private final PivnetCache pivnetCache;
+    private boolean spaceEventPublished;
```

Comment:
We are going to have to revisit this implementation and fix.  I know we talked through the design here where we wanted to account for the variability of configuration and the fact that we should wait on the results of two tasks when `om.enabled=true` and `pivnet.enabled=true`.

I'm thinking we should not have AppDetailTask implement ApplicationListener. And we should replace the existing onApplicationEvent method with multiple @EventListener annotated methods.  See this article: https://reflectoring.io/spring-boot-application-events-explained/.

We should remove the `Published`-suffixed `boolean` member variables and the `List<Space> spaces` member variable from this class as they make this class stateful, and that's not what we want.

So the last concern is... how can we account for collection when the act of invoking the collect method deletes all prior collected app detail records? We have a variable condition driven by configuration where we need to wait for the cache to be seeded before we can collect.

A new component will make this decision.  Be sure to inject into the constructor of `AppDetailTask`

```
@Component
public class AppDetailReadyDecider {

    private PivnetSettings settings; 
    private AtomicInteger decision = new AtomicInteger();
    private List<Space> spaces = new ArrayList<>();

    @Autowired
    public AppDetailReadyDecider(PivnetSettings settings) {
       this.settings = settings;
    }

    public int informDecision() {
      return decision.incrementAndGet();
    }

    public boolean isDecided() {
       return settings.isEnabled() ? decision.get() == 2: decision.get() == 1;
    }

  public void setSpaces(List<Space> spaces) {
    this.spaces = spaces;
  }

  public void reset() {
    spaces.clear();
    decision.set(0);
  }

  public List<Space> getSpaces() {
    return List.copyOf(spaces);
  }
```

```
  @EventListener
  void handleSpacesRetrieved(SpacesRetrievedEvent event) {
    decider.informDecision();
    decider.setSpaces(event.getSpaces());
  }

  @EventListener
  void handleProductsAndReleasesRetrieved(ProductsAndReleasesRetrievedEvent event) {
    decider.informDecision();
  }

  @EventListener
  void handleAppDetailReadyToBeRetrieved(AppDetailReadyToBeRetrievedEvent event) {
    if (decider.isDecided()) {
      collect(decider.getSpaces());
      decider.reset();
    } else {
      collect(List.copyOf(event.getSpaces()));
    }
  }
```